### ISSUE TITLE
Store config in alternative location (portable app)

### ISSUE THREAD

**Issue by pstein** [mifi/lossless-cut#645]
Where are settings of LC stored?
I can't see a file like losslesscut.ini in installation folder

So it must me either in Windows Registry or in User App folder.

Unfortunately LC is not portable this way.
I would like to run it from USB flash drive on various computers and without leaving traces on the computers.

So could you make it really portable?
The preferences should be stored in an *.ini or *.xml file in Losslesscut folder.
Thank you
Peter

**Comment by markusd1984** [comment]
This would make rollouts so much better as well, to set defaults in a .ini file, please.

**Comment by mifi** [comment]
It is pretty standard for apps these days to store their config inside the User folder so that many users can use the same app with diffferent configs, and you don't lose config if you delete the app (this is how it works on Linux, Mac and I believe also most Windows apps nowadays), because `c:/program files` should not be writeable by the user.

LosslessCut uses [electron-store](https://github.com/sindresorhus/electron-store) which uses [getAppPath](https://www.electronjs.org/docs/api/app#appgetpathname):

Per-user application data directory, which by default points to:
`%APPDATA%` on Windows
`$XDG_CONFIG_HOME` or `~/.config` on Linux
`~/Library/Application Support` on macOS

We can leave this open as a feature request but I'm not going to prioritize this anytime soon, because I don't know an easy way to do this (maybe it could be done with a command line flag (`-portable`), but that would be a bit cumbersome for users)

**Comment by markusd1984** [comment]
It makes absolutely sense why you used the popular user /platform default folders to store (user specific) app settings, we are just saying if we additionally could support using a local stored config file only when present (or take priority over present user setting), that this would make it **fully portable** as well as potentially easier for **rollouts** to multiple laptops. 

I use ShareX and Greenshot which use this, they also have full installers besides fully portable versions which is nice. We are currently building an installer via app deployment tool to deploy/make available via software center (SSCM), I will check out to utilise the current setting files, biggest hurdle is the lack of uninstall support due to no installer #521.)

Perhaps we could use a specific relative path via '[app.setPath(name,path)](https://www.electronjs.org/docs/api/app#appsetpathname-path)` as per https://stackoverflow.com/a/30610157/8719001 

(There is also a module called [electron-json-storage](https://github.com/electron-userland/electron-json-storage) in case that can be helpful, however a .ini file is more common to most people / editor friendly.)

**Comment by mifi** [comment]
Yea could use setPath but then we would need to know when to call that instead of the default. (It cannot be a setting for obvious reasons)
A command line flag could work

**Comment by mifi** [comment]
In addition, Electron internally uses the same path for storing a bunch if chrome caches and configs etc, i’m not sure if those would be affected by setAppPath too. If not then I’m pretty sure the app would crash if AppPath is not writable by the user

**Comment by markusd1984** [comment]
Can the path for the settings storage not be different for Chrome caches and other stuff? Unfortunately don't know anything about electron, just googling myself. :smile:

If yes I imagined that we could perhaps have a logic check to see if config.ini is present in the same folder as losslesscut.exe and if so use that for settings instead of user data directory, (otherwise fault back on default using user data directory).

**Comment by mifi** [comment]
If settings are still stored in User/AppData, then it would break the promise of not leaving traces on the computer.
If checking for config.ini and the file does not exist because the user forgot to create the file, then it would also leave traces on the computer

**Comment by pstein** [comment]
LC could check if a subfolder "settings" exist in the LC installation folder. If this subfolder exists then LC uses the files inside this folder otherwise is looks in the default User/AppData directory.

I don't know in which programming language LC is written. So in pseudo code it would look like:

If exists(getLCInstallationfolder()\settings) {
    settingsfolder=getLCInstallationfolder()\settings; }
else {
    settingsfolder=getAppDatafolder()\Roaming\Losslesscut; }

So making LC portable costs only 3 tiny lines of source code more.

Is this too difficult?

**Comment by markusd1984** [comment]
Shouldn't require a folder just a file in the same folder where the exe is to keep it simple.

Yes if electron still requires user/App Data folder that may not be fully portable or without trace, but this is happening currently regardless thus the main benefit is being able to deploy or use it portable **including preconfigured settings** . That's why we need it and should implement it.

Potentially it could be also a setting to enable local settings which creates the config.ini file and when present use that to populate all default settings and update with any changes, that would also make it user-friendly for changing settings as one just does the changes in the app and can copy or backup the config file. :smile:

**Comment by mifi** [comment]
I'm not going to spend too much time on this, but I will add a check for whether a config file exists in the app dir [getAppPath](https://www.electronjs.org/docs/api/app#appgetapppath)

**Comment by mattack1** [comment]
I know nothing about electron-store, but IMO it should be putting things in preferences files, not
~/Library/Application Support on macOS

though I guess if you have a whole bunch of different unrelated files, that's ok.. but I think of configuration info as largely the same as user preferences for an app.

**Comment by mifi** [comment]
I will only make the "portable" change for Windows. MacOs/Linux will always be using electron `app.getPath('appData')` (https://www.electronjs.org/docs/api/app#appgetpathname)
I believe most MacOs apps store configuration inside `~/Library/Application Support`, see for instance Chrome `~/Library/Application\ Support/Google/Chrome/Default`.
What's the benefit of storing under `~/Library/Preferences` ?

**Comment by mifi** [comment]
will be out in the next ver

**Comment by seiferflo** [comment]
> will be out in the next ver
Thanks, but how do we set it to portable mode on Windows since the 3.33.1 ? 
Data folder still appears in %APPDATA%\Roaming

**Comment by mifi** [comment]
Put an config.json file containing `{}` next to the exe and it should use that. It will still create electron stuff inside appdata/roaming, unfortunately i dont know if it’s possible to fix that

**Comment by tobwz** [comment]
How should users this know?

Wouldn't it be easier if you create a clickable checkbox/button/menu in GUI with the label "Run in portable mode" which prepare everything necessary?

**Comment by mifi** [comment]
either that or a command line flag to start it in portable mode can be done. I just don't think a lot of people need that, and if they do, then they can now still achieve it by creating this config file.

**Comment by seiferflo** [comment]
Thx for your feedback @mifi . I disagree in a sense that they are much more users than you think who prefer to have an app running in portable mode, simply to avoid leaving any trace or any leftovers in case they don't use it anymore. 
Keeping the app and its settings at the same place to be used simply on another computer is also important nowadays as people are very mobile.
So giving the choice to people of installing it or using it in portable as @tobwz mentioned is a good idea.

A neat way would to have an App & Data folder and a shortcut on the root level.
App contains what you have today with the exe.
Data contains what you create in the %APPDATA%\Roaming

Whether this is a achieved through a json file or portable command line flag is ok, as long as it's described somewhere. 
Speaking of which, creating a config.json containing what you mentioned doesn't seem to be doing anything in my case...

Thx for your consideration

**Comment by mifi** [comment]
> Thx for your feedback @mifi . I disagree in a sense that they are much more users than you think who prefer to have an app running in portable mode, simply to avoid leaving any trace or any leftovers in case they don't use it anymore.

The problem is that it is not that simple because lc is using a framework called electron which has some ways of doing things, like storing appdata.
I found this [app.setPath(name, path)](https://www.electronjs.org/docs/api/app#appsetpathname-path) however I think that would only work if we had a a command line flag (because it needs to run immediately upon startup, before app->ready event, I believe in the first tick) so we cannot run async code. (to check for an existing file)

> Speaking of which, creating a config.json containing what you mentioned doesn't seem to be doing anything in my case...

Are you running single EXE version? can you check developer tools if it logs `customStoragePath` somewhere there? or try to run the exe from the command line and see if it spits out the path there

**Comment by seiferflo** [comment]
Sorry I'm not understanding your first answer as I'm not a developer. 

For the exe, I'm using your zip version and there is only one LosslessCut.exe file in the root folder, where I put the config.json file there.

I've opened the developer tools and didn't find the customStoragePath log anywhere.

Launching the exe from cmd,gives me _"(electron) The default value of app.allowRendererProcessReuse is deprecated, it is currently "false".  It will change to be "true" in Electron 9."_

**Comment by mifi** [comment]
sorry those were just some notes for myself (and any other devs). In layman terms I believe it's not possible to have it store the other files (Electron app data) in the same dir as the exe file after detecting a config.json next to the exe because it is too late (electron has already stored the files in Roaming by then). if we want to store those files also next to exe, I think we need to run losslesscut with a command line flag that says to use exe dir for storage of those. but  then the user would have to start lc with a command line flag every time, i'm not sure how nice that is.

**Comment by mifi** [comment]
As for why config.json doesn't work, I have to look into that. I only tested on mac and here it worked, so i need to fire up a windows vm to test

**Comment by Mitch037** [comment]
Has there been any progress on the feature? I have noticed that placing config.json containing {} next to the exe doesn't do anything.

**Comment by mifi** [comment]
Are you running the self-contained LosslessCut-win.exe from github, e.g. this one? https://github.com/mifi/lossless-cut/releases/download/v3.39.0/LosslessCut-win.exe

I think that version is a self-extracting exe archive, and will extract all files into a temp directory like maybe `c:/Temp`, before running the extracted exe file from the temp file. I have a suspicion that `getAppPath` will return this temp folder instead.
What you could try is instead download the zip file, extract that to a directory and then make a config.json file inside there, if that fixes the problem, then I'm pretty sure that this is the issue. https://github.com/mifi/lossless-cut/releases/download/v3.39.0/LosslessCut-win.zip

Meanwhile I found this and will look into it:
https://stackoverflow.com/questions/46307797/how-to-get-the-original-path-of-a-portable-electron-app

**Comment by Mitch037** [comment]
I've been using the zip version from the very beginning.

**Comment by mifi** [comment]
Ok it seems there was a bug, where I used `getAppPath()`, but that [points to asar instead](https://github.com/electron/electron/issues/10975
). In the enxt version will instead use `getPath('exe')`, hopefully that fixed the problemn.

**Comment by JohnTHaller** [comment]
For folks who would like the Electron data in AppData handled as well, we've released [Lossless Cut Portable](https://portableapps.com/apps/music_video/lossless-cut-portable) at PortableApps.com. It uses the config.json method mentioned above as well as moving files to/from %APPDATA%\LosslessCut. It also adjusts paths within config.json as you move between computers.

**Comment by mifi** [comment]
Closing for now. I'll reopen if still an issue

**Comment by SoCuul** [comment]
> @mifi: I will only make the "portable" change for Windows. MacOs/Linux will always be using electron `app.getPath('appData')` (https://www.electronjs.org/docs/api/app#appgetpathname) I believe most MacOs apps store configuration inside `~/Library/Application Support`, see for instance Chrome `~/Library/Application\ Support/Google/Chrome/Default`. What's the benefit of storing under `~/Library/Preferences` ?

The `~/Library/Preferences` folder on macOS is akin to the registry on windows. Instead of storing the data together in a binary file, the convention is to have plist files with the name of the bundle id they are associated with, which is used by the native key/value store for macOS (NSUserDefaults). Eg. `com.google.Chrome.plist`.

It's typically better to store configuration data here (even in other file formats such as `.json`/`.ini`/`.xml`, as this leaves the user the ability to clear the application's data (in `~/Library/Application Support`), without losing user-specific settings. It will also be typically included as part of backups, even if general application data is excluded.

**Comment by bottleblue** [comment]
Is there a way to verify if the portable mode is working? I see the same folders in the App Data folder after putting the config.json file in the .exe folder.

**Comment by JohnTHaller** [comment]
> Is there a way to verify if the portable mode is working? I see the same folders in the App Data folder after putting the config.json file in the .exe folder.

The config.json is used for LosslessCut's own data. You'll still see Electron stuff within your local APPDATA (blob_storage, Cache, etc).

**Comment by subsubsubsubsubsub** [comment]
Why a portable version? If it's not portable? 
why not just make it portable? 
There's not even a warning that the portable is a fake for the sake of appearances.

**Comment by mifi** [comment]
Electron does not support portable mode afaik, and I don’t want to try to fit a square peg into a round hole.

**Comment by JohnTHaller** [comment]
You can use --user-data-dir="C:\ThePath\ToWhereYouWantIt" passed into LosslessCut.exe to redirect some of the Electron files to a location of your choice. Some data will still be saved into %APPDATA%\LosslessCut. The portable package at PortableApps.com makes use of that switch and then copies stuff back and forth to %APPDATA%\LosslessCut for the remainder. It also ensures that a portable copy of LosslessCut won't interfere with the settings of a local install.
